qiskit_ode/signals/signals.py

```diff
@@ -327,202 +407,570 @@ def add_samples(self, start_sample: int, samples: List):
+        )
 
-        Args:
-            t (float): time
+    def __call__(self, t: Union[float, np.array, Array]) -> Array:
+        """Vectorized evaluation of all components."""
+        return np.moveaxis(Array([Array(sig(t)).data for sig in self.components]), 0, -1)
 
-        Returns:
-            Array: the value of the signal (including carrier frequencies)
-                      at time t
-        """
-        carrier_val = np.exp(t * self._im_angular_freqs + 1.0j * self.phases)
-        return self.envelope_value(t) * carrier_val
+    def flatten(self) -> "SignalList":
+        """Return a ``SignalList`` with each component flattened."""
+        flattened_list = []
+        for sig in self.components:
+            if isinstance(sig, SignalSum):
+                flattened_list.append(sig.flatten())
+            else:
+                flattened_list.append(sig)
 
-    def conjugate(self):
-        """Return a new VectorSignal that is the complex conjugate of self.
+        return SignalList(flattened_list)
 
-        Returns:
-            VectorSignal: the complex conjugate of self
-        """
-        return VectorSignal(
-            lambda t: np.conjugate(self.envelope_value(t)),
-            -self.carrier_freqs,
-            -self.phases,
-            np.conjugate(self.drift_array),
+    @property
+    def drift(self) -> Array:
+        """Return the 'drift' Array, i.e. the constant part of the ``SignalList``."""
+
+        drift_array = []
+
+        for sig_entry in self.components:
+            val = 0.0
+
+            if not isinstance(sig_entry, SignalSum):
+                sig_entry = SignalSum(sig_entry)
+
+            for term in sig_entry:
+                if isinstance(term, Constant):
+                    val += term._value.data
+
+            drift_array.append(val)
+
+        return Array(drift_array)
+
+    def __getitem__(self, idx: Union[int, List, np.array, slice]) -> Signal:
+        """Enables numpy-style subscripting, as if this class were a 1d array."""
+
+        if type(idx) == np.ndarray and idx.ndim > 0:
+            idx = list(idx)
+
+        sublist = None
+
+        # get a list of the subcomponents
+        if type(idx) == list:
+            # handle lists
+            sublist = operator.itemgetter(*idx)(self.components)
+
+            # output will either be a single signal or a tuple of Signals
+            # convert to list if tuple
+            if type(sublist) == tuple:
+                sublist = list(sublist)
+        else:
+            # handle slices or singletons
+            sublist = operator.itemgetter(idx)(self.components)
+
+        # sublist should either be a single Signal, or a list of Signals
+        if isinstance(sublist, Signal):
+            return sublist
+        else:
+            return SignalList(*sublist)
+
+    def __len__(self):
+        return len(self.components)
+
+
+def signal_add(sig1: Signal, sig2: Signal) -> SignalSum:
```

Comment:
To keep things more modular it might make sense to put these type of functions in a `signals_operations.py` file.

Reply:
Just tried but got circular import issues, so I think we'll need to keep these together.